python/tskit/trees.py

```diff
@@ -5098,13 +5098,27 @@ def mutation(self, id_):
             metadata_decoder=self.table_metadata_schemas.mutation.decode_row,
         )
 
-    def site(self, id_):
+    def site(self, id_=None, position=None):
         """
         Returns the :ref:`site <sec_site_table_definition>` in this tree sequence
-        with the specified ID.
+        with either the specified ID or position.
 
         :rtype: :class:`Site`
         """
+        if id_ is None and position is None:
+            raise TypeError("Site id or position must be provided.")
+        if id_ is None:
+            if position < 0 or position >= self.sequence_length:
+                raise ValueError(
+                    "Position is beyond the coordinates defined by sequence length."
+                )
+            position = np.array(position)
+            if len(position.shape) > 0:
+                raise ValueError("Position must be provided as a scalar value.")
+            site_pos = self.tables.sites.position  # Sorted in ascending order
+            id_ = site_pos.searchsorted(position)
+            if site_pos[id_] != position:
```

Comment:
Don't you end up getting a `KeyError` here if the position is greater than the last site? `searchsorted` will return one-past-the-end.

Reply:
I have added this assertion in ``site`` before ``searchsorted``. It only works for discrete coordinates though.

```
if position < 0 or position >= self.sequence_length:
    raise ValueError("Position is beyond the coordinates defined by sequence length.")
```

Reply:
I don't think that works even for discrete coords - say for example if you had zero sites!

Reply:
I have added a test to check that a KeyError is returned in such an instance.

Reply:
I don't think `KeyError` is the right exception to pass back to the user here. Can you add a test for getting a site from a table with no sites also to exercise this part and check we get the right exception? (`ValueError`)

Reply:
I think I have corrected it now.